shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -317,6 +318,29 @@ - (NSOpenGLContext*)resourceContext {
   return _resourceContext;
 }
 
+- (float)displayRefreshRate {
+  NSDictionary* description = [[NSScreen mainScreen] deviceDescription];
+  CGDirectDisplayID displayId = [[description objectForKey:@"NSScreenNumber"] unsignedIntValue];
+  CGDisplayModeRef displayModeRef = CGDisplayCopyDisplayMode(displayId);
+  double refreshRate = CGDisplayModeGetRefreshRate(displayModeRef);
+
+  // The above can return zero always on some devices.
+  // See: https://github.com/glfw/glfw/issues/137
+  if (refreshRate == 0.0f) {
+    CVDisplayLinkRef displayLinkRef;
+    CVDisplayLinkCreateWithActiveCGDisplays(&displayLinkRef);
+
+    CVTime nominal = CVDisplayLinkGetNominalOutputVideoRefreshPeriod(displayLinkRef);
+    if (!(nominal.flags & kCVTimeIsIndefinite)) {
+      refreshRate = static_cast<double>(nominal.timeScale) / nominal.timeValue;
+    }
+
+    CVDisplayLinkRelease(displayLinkRef);
+  }
+
+  return static_cast<float>(refreshRate);
```

Comment:
Why are we using a float rather than a double?

Reply:
Related comments on #21095 as well for x-ref.

Reply:
The API that exposes the display refresh rate: https://github.com/flutter/engine/blob/master/shell/common/vsync_waiter.h#L37 returns a `float`. That was my motivation for returning a float.

Reply:
We can change engine internals at any time; once the embedder API is added we are stuck with it. Absent a very strong reason not to, we should use a double here, and if the engine wants to truncate it to a float it can do so.

Reply:
👍